community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/steps/verifyBestPlan.scala

```diff
@@ -19,15 +19,51 @@
  */
 package org.neo4j.cypher.internal.compiler.v2_3.planner.logical.steps
 
+import org.neo4j.cypher.internal.compiler.v2_3.IndexHintException
+import org.neo4j.cypher.internal.compiler.v2_3.ast.{LabelName, Identifier, UsingIndexHint}
+import org.neo4j.cypher.internal.compiler.v2_3.notification.IndexHintUnfulfillableNotification
 import org.neo4j.cypher.internal.compiler.v2_3.planner.logical.plans.LogicalPlan
 import org.neo4j.cypher.internal.compiler.v2_3.planner.logical.{LogicalPlanningContext, PlanTransformer}
 import org.neo4j.cypher.internal.compiler.v2_3.planner.{CantHandleQueryException, PlannerQuery}
+import org.neo4j.cypher.internal.compiler.v2_3.spi.PlanContext
 
 object verifyBestPlan extends PlanTransformer[PlannerQuery] {
   def apply(plan: LogicalPlan, expected: PlannerQuery)(implicit context: LogicalPlanningContext): LogicalPlan = {
     val constructed = plan.solved
-    if (expected != constructed)
-      throw new CantHandleQueryException(s"Expected \n$expected \n\n\nInstead, got: \n$constructed")
+    if (expected != constructed) {
+      val unfulfillableHints = unfulfillableIndexHints(expected, context.planContext)
+      val expectedWithoutHints = expected.withoutHints(unfulfillableHints)
+      if (expectedWithoutHints != constructed) {
+        if (expected.withoutHints(expected.allHints) != constructed.withoutHints(constructed.allHints)) {
+          // unknown planner issue failed to find plan (without regard for differences in hints)
+          throw new CantHandleQueryException(s"Expected \n$expected \n\n\nInstead, got: \n$constructed")
+        } else {
+          // unknown planner issue failed to find plan matching hints (i.e. "implicit hints")
+          throw new CantHandleQueryException(s"Expected \n${expected.allHints} \n\n\nInstead, got: \n${constructed.allHints}")
+        }
+      } else {
+        // hints referred to non-existent indexes ("explicit hints")
+        if (context.useErrorsOverWarnings) {
+          val firstIndexHint = unfulfillableHints.head
+          throw new IndexHintException(firstIndexHint.identifier.name, firstIndexHint.label.name, firstIndexHint.property.name, "No such index")
+        } else {
+          context.notificationLogger.log(IndexHintUnfulfillableNotification)
```

Comment:
There should be one notification per hint and it should have more details, shouldn't it?